ogre2/src/Ogre2COMVisual.cc

```diff
@@ -0,0 +1,218 @@
+ *
+ */
+
+#include "ignition/rendering/ogre2/Ogre2COMVisual.hh"
+#include "ignition/rendering/ogre2/Ogre2DynamicRenderable.hh"
+#include "ignition/rendering/ogre2/Ogre2Material.hh"
+
+#ifdef _MSC_VER
+  #pragma warning(push, 0)
+#endif
+#include <OgreSceneNode.h>
+#ifdef _MSC_VER
+  #pragma warning(pop)
+#endif
+
+class ignition::rendering::Ogre2COMVisualPrivate
+{
+  /// \brief Grid materal
+  public: Ogre2MaterialPtr material = nullptr;
+
+  /// \brief Lines that make the cross marking the center of mass.
+  public: std::shared_ptr<Ogre2DynamicRenderable> crossLines = nullptr;
+
+  /// \brief Sphere visual marking the center of mass
+  public: VisualPtr sphereVis = nullptr;
+};
+
+using namespace ignition;
+using namespace rendering;
+
+//////////////////////////////////////////////////
+Ogre2COMVisual::Ogre2COMVisual()
+  : dataPtr(new Ogre2COMVisualPrivate)
+{
+}
+
+//////////////////////////////////////////////////
+Ogre2COMVisual::~Ogre2COMVisual()
+{
+}
+
+//////////////////////////////////////////////////
+void Ogre2COMVisual::PreRender()
+{
+  if (this->dirtyCOMVisual)
+  {
+    this->CreateVisual();
+    this->dirtyCOMVisual = false;
+  }
+}
+
+//////////////////////////////////////////////////
+void Ogre2COMVisual::Init()
+{
+  BaseCOMVisual::Init();
+}
+
+//////////////////////////////////////////////////
+void Ogre2COMVisual::Destroy()
+{
+  if (this->dataPtr->sphereVis != nullptr)
+  {
+    this->dataPtr->sphereVis->Destroy();
+    this->dataPtr->sphereVis.reset();
+  }
+
+  if (this->dataPtr->crossLines)
+  {
+    this->dataPtr->crossLines->Destroy();
+    this->dataPtr->crossLines.reset();
+  }
+
+  if (this->dataPtr->material && this->Scene())
+  {
+    this->Scene()->DestroyMaterial(this->dataPtr->material);
+    this->dataPtr->material.reset();
+  }
+}
+
+//////////////////////////////////////////////////
+void Ogre2COMVisual::CreateVisual()
+{
+  if (!this->dataPtr->crossLines)
+  {
+    this->dataPtr->crossLines.reset(
+        new Ogre2DynamicRenderable(this->Scene()));
+    this->ogreNode->attachObject(this->dataPtr->crossLines->OgreObject());
+  }
+
+  if (!this->dataPtr->sphereVis)
+  {
+    this->dataPtr->sphereVis = this->Scene()->CreateVisual();
+    this->dataPtr->sphereVis->AddGeometry(this->Scene()->CreateSphere());
+    this->dataPtr->sphereVis->SetMaterial("Default/CoM");
+    this->dataPtr->sphereVis->SetInheritScale(false);
+    this->AddChild(this->dataPtr->sphereVis);
+  }
+
+  // Compute radius of sphere with density of lead and equivalent mass.
+  double sphereRadius;
+  double dLead = 11340;
+  sphereRadius = cbrt((0.75 * this->Mass()) / (IGN_PI * dLead));
+
+  this->dataPtr->sphereVis->SetLocalScale(ignition::math::Vector3d(
+      sphereRadius*2, sphereRadius*2, sphereRadius*2));
+  this->dataPtr->sphereVis->SetLocalPosition(this->InertiaPose().Pos());
+  this->dataPtr->sphereVis->SetLocalRotation(this->InertiaPose().Rot());
+
+  // Get the link's bounding box
+  if (this->ParentLink().empty() ||
```

Comment:
I would like to avoid the need for users to call `SetParentLink` function for setting `linkName` because we should be able to access the parent once the COMVisual is attached to the parent visual.

Once idea is to override the `PreRender` function:
* check to see if `HasParent()` is true and parent name is empty in `PreRender`, 
* then Update `crossLines` based on the parent's bounding box. 

Note that we should also avoid using the term `link` as that's ign-gazebo specific. I think `parentName` is more suitable.

Reply:
Fixed in [`f428ba1`](https://github.com/ignitionrobotics/ign-rendering/pull/345/commits/f428ba1a17054dee412110e293f2868f74c68f96). I have changed the API to use `parentName` without changing the desired behavior. Also, I think the Ogre1 `BoundingBox` issue (https://github.com/ignitionrobotics/ign-rendering/pull/345#issuecomment-863188351) is probably not related to this PR. It's working fine for me in Ogre2.